File: netty-on-freebsd/src/main/scala/so/chaos/demo/server/EchoServerHandler.scala
package so.chaos.demo.server

import io.netty.buffer.{ByteBuf, Unpooled}
import io.netty.channel.{ChannelFutureListener, ChannelHandler, ChannelHandlerContext, ChannelInboundHandlerAdapter}
import io.netty.util.CharsetUtil
import io.netty.util.internal.StringUtil
import org.slf4j.LoggerFactory

/**
  * <p>
  * <strong>
  * 用一句话描述功能
  * </strong><br /><br />
  * 如题。
  * </p>
  *
  * @author chengchaos[as]Administrator - 2019/5/14 0014 上午 11:34 <br />
  * @see 【相关类方法】
  * @since 1.1.0
  */
@ChannelHandler.Sharable
class EchoServerHandler extends ChannelInboundHandlerAdapter{

  private val logger = LoggerFactory.getLogger(classOf[EchoServerHandler])

  override def channelRead(ctx: ChannelHandlerContext, msg: Any): Unit = {
    val in = msg.asInstanceOf[ByteBuf]

    if (logger.isInfoEnabled) {
      logger.info("Server received ==> {}", in.toString(CharsetUtil.UTF_8))
    }

    ctx.write(in)
  }

  override def channelReadComplete(ctx: _root_.io.netty.channel.ChannelHandlerContext): Unit = {
    ctx.writeAndFlush(Unpooled.EMPTY_BUFFER)
      .addListener(ChannelFutureListener.CLOSE)
  }

  override def exceptionCaught(ctx: _root_.io.netty.channel.ChannelHandlerContext, cause: _root_.java.lang.Throwable): Unit = {
    logger.error(StringUtil.EMPTY_STRING, cause)
    ctx.close
  }
}

object EchoServerHandler {

  def apply(): EchoServerHandler = new EchoServerHandler()
  val INSTANCE = EchoServerHandler()

}

File: group-netty/my-netty-001/src/main/java/com/example/myscala002/netty/EchoServerhandler.java
package com.example.myscala002.netty;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.channel.ChannelFutureListener;
import io.netty.channel.ChannelHandler;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.util.CharsetUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * <p>
 * <strong>
 * 用一句话描述功能
 * </strong><br /><br />
 * 功能的详细描述
 * </p>
 *
 * @author chengchaos[as]Administrator - 2019/4/4 0004 下午 3:34 <br />
 * @see 【相关类方法】
 * @since 1.1.0
 */
@ChannelHandler.Sharable
public class EchoServerhandler extends ChannelInboundHandlerAdapter {

    private static final Logger LOGGER = LoggerFactory.getLogger(EchoServerhandler.class);


    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {

        ByteBuf in = (ByteBuf) msg;

        if (LOGGER.isInfoEnabled()) {
            LOGGER.info("Server received ==> {}", in.toString(CharsetUtil.UTF_8));
        }

        ctx.write(in);

    }

    @Override
    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {

        ctx.writeAndFlush(Unpooled.EMPTY_BUFFER)
                .addListener(ChannelFutureListener.CLOSE);
    }

    /**
     * 记录异常，关闭 Channel
     *
     * @param ctx
     * @param cause
     * @throws Exception
     */
    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {

        LOGGER.error("", cause);
        ctx.close();
    }
}
